pixy/consumer.go

```diff
@@ -0,0 +1,543 @@
+package pixy
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/mailgun/kafka-pixy/Godeps/_workspace/src/github.com/mailgun/kazoo-go"
+	"github.com/mailgun/kafka-pixy/Godeps/_workspace/src/github.com/mailgun/log"
+	"github.com/mailgun/kafka-pixy/Godeps/_workspace/src/github.com/mailgun/sarama"
+)
+
+type ErrConsumerSetup error
+type ErrConsumerBufferOverflow error
+type ErrConsumerRequestTimeout error
+
+// SmartConsumer is a Kafka consumer implementation that automatically
+// maintains consumer groups registrations and topic subscriptions. Whenever a
+// a message from a particular topic is consumed by a particular consumer group
+// SmartConsumer checks if it has registered with the consumer group, and
+// registers otherwise. Then it checks if it has subscribed for the topic, and
+// subscribes otherwise. Later if a particular topic has not been consumed for
+// the `Config.Consumer.RegistrationTimeout` period of time, then the consumer
+// unsubscribes from the topic, likewise if a consumer group has not seen any
+// requests for that period then the consumer deregisters from the group.
+//
+// Note that during state transitions subscribe<->unsubscribe and
+// register<->deregister the consumer may return either `ErrConsumerRequestTimeout`
+// or `ErrConsumerBufferOverflow` error. In that case the user should back off
```

Comment:
I assume you mean that `Consume` may return those errors. Might be helpful to say that explicitly.